proposals/js-promise-integration/Overview.md

```diff
@@ -91,210 +66,147 @@ WebAssembly (`demo.wasm`):
 
-The `Suspender` object specified here is not made available to JavaScript via this API. Unless exported via some form of back-channel it will not participate in normal JavaScript execution. However, it does have an internal role within the API and so it is specified.
+The `WebAssembly.promising` function takes a WebAssembly function -- i.e., not a JavaScript function -- and converts it to a JavaScript function that returns a `Promise`.
 
-A `Suspender` is in one of the following states:
-* **Moribund** - not available for use.
-* **Active**[`caller`] - control is inside the `Suspender`, with `caller` being the function that called into the `Suspender` and is expecting an `externref` to be returned
-* **Suspended** - currently waiting for some promise to resolve
+### Suspendable functions
 
-Note that within a WebAssembly module, a `Suspender` is typed as an `externref`.
+>In the following description we use the term *execution context* to denote a potentially suspendable computation. This should not be confused with other uses of the term.
 
-### Suspending Functions
+We modify the *read-the-imports* algorithm in the WebAssembly [JS-API](https://webassembly.github.io/spec/js-api/index.html#read-the-imports) specification to account for functions marked as `Suspending`. In particular, we add the clause:
 
-The constructor for `WebAssembly.Function`, when it has a `suspending` attribute in its `usage` dictionary:
+1. Let *`o`* be `$Get$`(*`importObject`*, *`moduleName`*).
+1. If *`o`* is of the form *Suspendable*,
+  1.  Let *`v`* be  `$Get$`(*`o`*, *`wrappedFunction`*).
+    1. If `$IsCallable$`(*`v`*) is false, throw a `LinkError` exception.
+      1. Create a *suspending function* from *`v`* and *`functype`*, and let *`funcaddr`* be the result.
+  1. Let *`externfunc`* be the external value *`funcaddr`*
+  1. Append *`externfunc`* to *`imports`*.
 
-* If the value of the `suspending` attribute is `"first"`, the `type` argument of the constructor must be of the form:
-  ```
-  { parameters: ["externref", t0, .., tn], results: [r0, .., rk]}
-  ```
-* If the value of the `suspending` attrbute is `"last"`:
-  ```
-  { parameters: [t0, .., tn, "externref"], results: [r0, .., rk]}
-  ```
-* If the value of the `suspending` attribute is `"none"`:
-  ```
-  { parameters: [t0, .., tn], results: [r0, .., rk]}
-  ```
-The WebAssembly function returned by `WebAssembly.Function` is a marked function whose behavior is determined as follows:
+The *suspending function* is a function whose behavior is determined as follows:
 
-0. Let `suspender` be an additional argument that is expected to contain a `Suspender` object (with WebAssembly type `externref`). Let `func` be the wrapped function that was used when creating the `WebAssembly.Function`. 
+1. Let `context` refer to the execution context that is current at the time of a call to the *suspending function*. Let `func` be the wrapped function that was used when creating the *suspending function*.
```

Comment:
It seems like there is some detail missing defining this "execution context" concept in terms of existing Core WebAssembly and JS API spec concepts.  In particular, I was looking for something delimited by the innermost call into core wasm from the JS API.  As is, I'm not sure what the spec-level "type" of `context` is.

Reply:
Could be. However, 'innermost' is replaced by the explicit split point on the call to the export + prevention of calls to JS.

Reply:
The term 'execution context' is actually an EcmaScript concept; one that appears to align very neatly with what we are attempting to achieve: there it represents suspendable computations mediated by Promises.